### ISSUE TITLE
Question: Do I need to use a context manager with all ClientSession HTTP methods?

### ISSUE THREAD

**Issue by vimusov** [aio-libs/aiohttp#2154]
Hello!

Let's assume that I need to make a POST-request which returns nothing. Do I need to use a context manager and write a code like that?

```python
session = ClientSession(raise_for_status=True)  # one session for all requests!

async with session.post(...) as response:
    pass  # we already set `raise_for_status' so there is no need to call the method
```

Or I can just write:

```python
await session.post(...)
```

Which choice is better and more correctly?

**Comment by socketpair** [comment]
http://aiohttp.readthedocs.io/en/stable/client_reference.html#aiohttp.ClientResponse.release :

It is not required to call release on the response object. When the client fully receives the payload, the underlying connection automatically returns back to pool. If the payload is not fully read, the connection is closed

It's a bug. We should define some constant which say how many bytes we should internally read (and discard) before we give up, log warning (!) and close the connection. Typically, data sent in response with code 200 is not ever used, so code above according to current implementation will prevent usage of keep-alive either with `with` or without (since we do not read payload).

**Comment by fafhrd91** [comment]
aiohttp can not make decision on this, if payload is not completely readout, connection is in broken state. 
this is application concern.

**Comment by socketpair** [comment]
I don't understand why. If user did not read content -- this mean content is not important for him. Why not to remove it from socket and discard ? On other hand, big responses consumes CPU and make connection useless. That's why we need to define a border after which reading trash becomes useless. It's better to read a few bytes instead of require to re-open connection.

**Comment by fafhrd91** [comment]
If data is in buffer then it get discarded and connection get returned back to pool. But active reading from socket is application task.

`await rest.read()` has same effect as `release` from aiohttp 1.x

**Comment by fafhrd91** [comment]
Actually, connection get returned back to pool immidietly as http message eof received. Even if application code continues to work with response data

**Comment by socketpair** [comment]
So, why connection will be closed if response is not read ? (according to documentation).
Do docs mean immediate closing ?

It seems we do not understand each other.

**Comment by fafhrd91** [comment]
I see, doc should be modified. here `response` is actual data from socket, not data in aiohttp stream object.

**Comment by socketpair** [comment]
Well, I do not understand the difference. Will ask details which maybe give me some hint how the things work: What if response to POST request will not be read (as show in examples) and response will be 2 TB sent with speed of 32 kbit/s ? what will happen (in details) ? Sure it's very important for me in order to write efficient web applications based on aiohttp.

**Comment by fafhrd91** [comment]
in your case connection will be closed. 
but if your post response is 1kb, and 512bytes are headers and you get it as one tcp packet then
even if you do not read payload connection will be reused.

**Comment by socketpair** [comment]
Is it good to add some argument to `ClientSession()` and/or `.request()` `max_prefetch_response` which will force aiohttp to prefetch into internal buffer no more than specified count of bytes of response payload BEFORE resolving `.request()` coroutine ?

**Comment by fafhrd91** [comment]
I am not sure about prefetching. you can just wait on read

**Comment by socketpair** [comment]
I'm speaking about the case when aiohttp user does not read payload at all. It's not mistake and it should not affect keep-alive logic. So waiting on read is not a thing I speak about. Or I did not understand you again.

**Comment by fafhrd91** [comment]
it looks like wrong solution.

@asvetlov @kxepal thoughts?

**Comment by fafhrd91** [comment]
@socketpair question, is it that hard to read response before releasing?

**Comment by socketpair** [comment]
It's not hard at all :)

But it is not obvious for users. They will forget about this. This require (us) to instruct our internal linter which will check for that mistake.

The similar not-a-mistake is now fixed by `ClientSession(raise_for_status=True)`. Before that so called "hiding of error" took place and real error was not popped up.

The same here. But actually problem is much lighter, but still exists. An user expects one obvious behavior, but in practice things are going in a different way.

**Comment by fafhrd91** [comment]
I see. but I still don't like prefetching. 

maybe special context manager?

**Comment by socketpair** [comment]
I will think a bit later.

But I don't like idea with one more level of nesting code.

**Comment by fafhrd91** [comment]
I don't like prefetching because feature as a name "prefetch" suggest is pretty useless and real intention is different.

**Comment by kxepal** [comment]
I'm actually don't follow the topic you're fighting for (:
Trying to summarize for myself:
1. You made a request via `resp = await session.post()`
2. According current logic the part of response got read to let you access to headers without any else async operation.
3. From here response awaits (literaly) for you next instructions: will you read the payload or not.
4. If you don't, connection still keeps alive due to TCP and actually, I not sure if some part of payload didn't get buffered on some low level meanwhile. Because server doesn't know what are you doing, it keep sending the data until your socket recvbuf won't be full.

Why should half-readed response be closed by aiohttp automagically and how? Only by some sort of read-timeout, but hey, if you don't need a response payload, just call `.release()` explicitly and cary on. If you read a streamed payload and in the middle decided that the rest part of it is not interested to you, do the same. 

Imagine that instead of reading responses, you're reading files. You opened a file, read a part and didn't explicitly closed. And another one. And again. What happens? Your open fds number counting until hit the limit. The only way to save yourself here from crash is close the files while you don't need them anymore. With the responses there is the same story, expect you can release them to return connection back to the pool.

So, assume reponse as a async file. Should you explicitly release/close it when you don't need it? Yes. Should aiohttp implement own GC to cleanup after you? No. For a single line of code you forgot to write it's too epic task.

Did I get you right, folks?

**Comment by fafhrd91** [comment]
problem is `release()` call semantics. in aiohttp 1.x it consumes whole payload, in 2.x it does not and technically useless. reason for change was backward compatibility and because we always suggested to do "release" for each response, which is not what most people wants.

maybe we should restore `release` behavior?

**Comment by kxepal** [comment]
> maybe we should restore release behavior?

I'm fine with it. Explicit behaviour is good.

**Comment by fafhrd91** [comment]
-1 for background read. everything has to be explicit, developer has to make this decision.

for your case, you can just wrap ClientSession and do prefetching in your code. we can modify aiohttp to make it easier.

**Comment by kxepal** [comment]
Can we do something like GC-based release? If response object is used no longer, we do release before it get vanished. Is that possible?

Also +1 for -1 background read.  That's may hurt when you don't expect that.

**Comment by fafhrd91** [comment]
there is no reason to do GC. response and connection objects are decoupled, whenever response receives all data for http message connection get released. so if response still holds connection during `del` then connection is in dirty state and can not be reused. to make connection clean user has to read whole http message. 

but @socketpair has different problem, he wants read-out response automatically, so `release` step could be skipped.

**Comment by socketpair** [comment]
Also, `ClientSession(raise_for_status=True)` wil not read error message until the end.

**Comment by asvetlov** [comment]
Closing, feel free to reopen/create a new issue if needed.
